src/lib/storage/table.cpp

```diff
@@ -16,54 +16,66 @@
 
 void Table::add_column(const std::string& name, const std::string& type) {
-  // Implementation goes here
+  DebugAssert(row_count() == 0, "The table already contains data so adding a column is not possible.");
+
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+
+  auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(type);
+
+  _chunks.front()->add_segment(segment);
 }
 
 void Table::append(std::vector<AllTypeVariant> values) {
-  // Implementation goes here
-}
+  DebugAssert(values.size() == column_count(), "The number of values doesn't match number of columns.");
 
-uint16_t Table::column_count() const {
-  // Implementation goes here
-  return 0;
+  if (_chunks.back()->size() >= _max_chunk_size) {
+    _chunks.push_back(std::make_shared<Chunk>());
+
+    for (const auto& _column_type : _column_types) {
+      auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(_column_type);
+
+      _chunks.back()->add_segment(segment);
+    }
+  }
+
+  _chunks.back()->append(values);
 }
 
+uint16_t Table::column_count() const { return _chunks.front()->column_count(); }
+
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  uint64_t number_of_rows = 0;
+  for (const auto& chunk : _chunks) {
+    number_of_rows += chunk->size();
+  }
+  return number_of_rows;
 }
 
-ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return (ChunkID)_chunks.size(); }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
-}
+  auto column_id =
+      std::distance(_column_names.cbegin(), std::find(_column_names.cbegin(), _column_names.cend(), column_name));
 
-uint32_t Table::max_chunk_size() const {
-  // Implementation goes here
-  return 0;
-}
+  DebugAssert(column_id < static_cast<int64_t>(_column_names.size()), "The column couldn't be found.");
 
-const std::vector<std::string>& Table::column_names() const {
-  throw std::runtime_error("Implement Table::column_names()");
+  return (ColumnID)column_id;
 }
 
-const std::string& Table::column_name(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_name");
-}
+uint32_t Table::max_chunk_size() const { return _max_chunk_size; }
 
-const std::string& Table::column_type(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_type");
-}
+const std::vector<std::string>& Table::column_names() const { return _column_names; }
+
+const std::string& Table::column_name(ColumnID column_id) const { return _column_names.at(column_id); }
+
+const std::string& Table::column_type(ColumnID column_id) const { return _column_types.at(column_id); }
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::get_chunk"); }
+Chunk& Table::get_chunk(ChunkID chunk_id) { return *(_chunks.at(chunk_id)); }
```

Comment:
```return reinterpret_cast<Chunk &>(_chunks.at(chunk_id));```
Looks more C++ - like to me :)

Reply:
How about holding Chunks in the std::vector instead of shared pointers as a Chunk itself is just a vector of shared pointers (to BaseSegment) already. 
Then you can just do
```
  DebugAssert(chunk_id < m_chunks.size(), "Assert in Table::get_chunk(): There is no chunk for the given chunk_id");
  return m_chunks[chunk_id];
```